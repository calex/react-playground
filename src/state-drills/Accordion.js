import React from 'react';

class Accordion extends React.Component {
    static defaultProps = {
        sections: []
    }

    state = {
        accordionActiveIndex: 0
    };

    toggleAccordionState = (indexNumber) => {
        this.setState({
            accordionActiveIndex: (this.state.accordionActiveIndex === indexNumber) ? null : indexNumber
        })
    }

    renderContent(activeDisplayIndex) {
        const currentSection = this.props.sections[activeDisplayIndex]
        return (
            <p className='content'>
                {currentSection.content}
            </p>
        )
    }

    renderListItems() {
        return this.props.sections.map((section, index) => (
            <li key={index}>
                <button onClick={() => this.toggleAccordionState(index)}>
                    {section.title}
                </button>
                {(index === this.state.accordionActiveIndex) && this.renderContent(this.state.accordionActiveIndex)}
            </li>
        ))
    }

    render() {
        return (
            <ul>
                {!!this.props.sections.length && this.renderListItems()}
            </ul>
        )
    }
}

export default Accordion;