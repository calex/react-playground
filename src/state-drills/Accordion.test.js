import React from 'react';

import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Accordion from './Accordion.js';

describe(`Accordion Component`, () => {

  const sections = [
    {
        title: 'Section 1',
        content: 'Section 1 Lorem ipsum dolor sit amet consectetur adipisicing elit.',
    },
    {
        title: 'Section 2',
        content: 'Section 2 Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!',
    },
    {
        title: 'Section 3',
        content: 'Section 3 Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?',
    },
  ]
  
  it('renders empty given no accordion item data', () => {
    const wrapper = shallow(<Accordion />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders no active sections by default', () => {
    const wrapper = shallow(<Accordion sections={sections} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('shows content under a button on click', () => {
    const wrapper = shallow(<Accordion sections={sections} />)
    wrapper.find('button').at(1).simulate('click')
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('hides content under a button on double click', () => {
    const wrapper = shallow(<Accordion sections={sections} />)
    wrapper.find('button').at(1).simulate('click')
    wrapper.find('button').at(1).simulate('click')
    expect(toJson(wrapper)).toMatchSnapshot()
  })
  
  
  it('shows content under a button on click after another button is exposed', () => {
    const wrapper = shallow(<Accordion sections={sections} />)
    wrapper.find('button').at(1).simulate('click')
    wrapper.find('button').at(2).simulate('click')
    expect(toJson(wrapper)).toMatchSnapshot()
  }) 
})
