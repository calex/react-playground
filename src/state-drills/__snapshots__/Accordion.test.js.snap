// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion Component hides content under a button on double click 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component renders empty given no accordion item data 1`] = `<ul />`;

exports[`Accordion Component renders no active sections by default 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
    <p
      className="content"
    >
      Section 1 Lorem ipsum dolor sit amet consectetur adipisicing elit.
    </p>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component shows content under a button on click 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
    <p
      className="content"
    >
      Section 2 Cupiditate tenetur aliquam necessitatibus id distinctio quas nihil ipsam nisi modi!
    </p>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
  </li>
</ul>
`;

exports[`Accordion Component shows content under a button on click after another button is exposed 1`] = `
<ul>
  <li
    key="0"
  >
    <button
      onClick={[Function]}
    >
      Section 1
    </button>
  </li>
  <li
    key="1"
  >
    <button
      onClick={[Function]}
    >
      Section 2
    </button>
  </li>
  <li
    key="2"
  >
    <button
      onClick={[Function]}
    >
      Section 3
    </button>
    <p
      className="content"
    >
      Section 3 Animi amet cumque sint cupiditate officia ab voluptatibus libero optio et?
    </p>
  </li>
</ul>
`;
